apps/settings/main.py

```diff
@@ -1,236 +1,158 @@
+            "systemctl restart zpui.service"
+        )
+        self.steps.append(restart_service)
+
+        clean_tmp_dir = UpdateStep("cleaning tmp dir", lambda: shutil.rmtree(tmp_dir))
+        self.steps.append(clean_tmp_dir)
+
+
+class UpdateStep(object):
+    def __init__(self, name, do=None, undo=None, accepted_return_code=None):
+        self.accepted_return_code = accepted_return_code
+        self.name = name
+        self._do = do
+        self._undo = undo
+        self.return_value = None
+
+    def do(self):
+        if not self._do:
+            return
+        logger.debug("running '{}'".format(self.name))
+        if isinstance(self._do, basestring):
+            print(shlex.split(self._do))
+            self.return_value = subprocess.call(shlex.split(self._do))
         else:
-            logger.info("Update successful!")
-            sleep(0.5)  # showing the completed progressbar
-            pb.stop()
-            PrettyPrinter("Update successful!", i, o, 3)
-            self.suggest_restart()
-
-    def suggest_restart(self):
-        needs_restart = DialogBox('yn', i, o, message="Restart ZPUI?").activate()
-        if needs_restart:
-            os.kill(os.getpid(), signal.SIGTERM)
-
-
-class GitUpdater(GenericUpdater):
-    branch = "master"
-
-    steps = ["check_connection", "check_git", "check_revisions", "pull", "install_requirements", "tests"]
-    progressbar_messages = {
-        "check_connection": "Connection check",
-        "check_git": "Running git",
-        "check_revisions": "Comparing code",
-        "pull": "Fetching code",
-        "install_requirements": "Installing packages",
-        "tests": "Running tests",
-    }
-    failed_messages = {
-        "check_connection": "No Internet connection!",
-        "check_git": "Git binary not found!",
-        "check_revisions": "Exception while comparing revisions!",
-        "pull": "Couldn't get new code!",
-        "install_requirements": "Failed to install new packages!",
-        "tests": "Tests failed!"
-    }
-
-    def do_check_git(self):
-        if not GitInterface.git_available():
-            logger.exception("Couldn't execute git - not found?")
-            raise OSError()
-
-    def do_check_revisions(self):
-        GitInterface.command("fetch")
-        current_branch_name = GitInterface.get_current_branch()
-        current_revision = GitInterface.get_head_for_branch(current_branch_name)
-        remote_revision = GitInterface.get_head_for_branch("origin/"+current_branch_name)
-        if current_revision == remote_revision:
-            raise UpdateUnnecessary
+            self.return_value = self._do()
+        logger.debug("ran '{}' with return value '{}'".format(self.name, self.return_value))
+        if self.accepted_return_code is not None:
+            if not self.return_value == self.accepted_return_code:
+                raise Exception("'{}' failed".format(self.name))
+        return self.return_value
+
+    def undo(self):
```

Comment:
```
[ERROR] 2017-12-28 10:05:45 main.py [26565]: Error updating step 'running tests'
[ERROR] 2017-12-28 10:05:45 main.py [26565]: 'running tests' failed
Traceback (most recent call last):
  File "/root/ZPUI/apps/settings/main.py", line 53, in start_update
    step.do()
  File "/root/ZPUI/apps/settings/main.py", line 150, in do
    raise Exception("'{}' failed".format(self.name))
Exception: 'running tests' failed
[WARNING] 2017-12-28 10:05:47 main.py [26565]: no undo method for 'copying repo'
[WARNING] 2017-12-28 10:05:47 main.py [26565]: no undo method for 'pulling git'
[ERROR] 2017-12-28 10:05:48 input.py [26565]: Exception 'str' object is not callable caused by
 callback <function <lambda> at 0xb50e1df0> when key KEY_ENTER was received
[ERROR] 2017-12-28 10:05:48 input.py [26565]: Traceback (most recent call last):
  File "/root/ZPUI/input/input.py", line 179, in handle_callback
    callback()
  File "/root/ZPUI/ui/base_list_ui.py", line 278, in <lambda>
    "KEY_ENTER":lambda: self.select_entry()
  File "/root/ZPUI/ui/utils.py", line 16, in wrapper
    return func(self, *args, **kwargs)
  File "/root/ZPUI/ui/menu.py", line 85, in select_entry
    entry[1]()
  File "/root/ZPUI/apps/settings/main.py", line 38, in update_zpui
    self.start_update(branch="master")
  File "/root/ZPUI/apps/settings/main.py", line 61, in start_update
    self.rollback_update(bar, e, finished_steps, step)
  File "/root/ZPUI/apps/settings/main.py", line 72, in rollback_update
    passed.undo()
  File "/root/ZPUI/apps/settings/main.py", line 158, in undo
    self._undo()
TypeError: 'str' object is not callable

[ERROR] 2017-12-28 10:05:48 input.py [26565]: Locals:
[ERROR] 2017-12-28 10:05:48 input.py [26565]: {'self': <apps.settings.main.UpdateStep object a
t 0xb5071d90>}
```
This was the last problem I've encountered, and it completely stopped the update process for me - I had to Ctrl+C out of it (which won't be possible for an user that doesn't have access to the console and is trying to update from within ZPUI).